os/pm/pm_metrics.c

```diff
@@ -17,117 +17,263 @@
+		((double)g_pm_metrics->idle_suspend_counts[index]) * 100.0 / ((double)g_pm_metrics->total_suspend_counts));
+	}
+	pmdbg("\n\n");
+	if (total_wakeup_src_counts) {
+		pmdbg(" WAKEUP SOURCES |      COUNTS      \n");
+		pmdbg("----------------|------------------\n");
+		for (index = 0; index < PM_WAKEUP_SRC_COUNT; index++) {
+			pmdbg(" %14s | %6d (%6.2f%%) \n", wakeup_src_name[index], g_pm_metrics->wakeup_src_counts[index],\
+			((double)g_pm_metrics->wakeup_src_counts[index]) * 100.0 / total_wakeup_src_counts);
+		}
+	}
+	pmdbg("\n\n");
+	pmdbg("    STATE    |          TIME          \n");
+	pmdbg("-------------|------------------------\n");
+	for (index = 0; index < PM_COUNT; index++) {
+		pmdbg(" %11s | %10dms (%6.2f%%) \n", state_name[index], TICK2MSEC(g_pm_metrics->sticks[index]),\
+		((double)g_pm_metrics->sticks[index]) * 100.0 / total_time);
+	}
+	pmdbg("\n\n");
+	pmdbg("    STATE    |          TIME          \n");
+	pmdbg("-------------|------------------------\n");
+	pmdbg(" %11s | %10dms (%6.2f%%) \n", "SLEEP", TICK2MSEC(g_pm_metrics->board_sleep_time),\
+	((double)g_pm_metrics->board_sleep_time) * 100.0 / total_time);
+	pmdbg(" %11s | %10dms (%6.2f%%) \n", "WAKEUP",((int)total_time) - TICK2MSEC(g_pm_metrics->board_sleep_time),\
+	(total_time - ((double)g_pm_metrics->board_sleep_time)) * 100.0 / total_time);
+}
+/************************************************************************************
+ * Public Functions
+ ************************************************************************************/
 
+/****************************************************************************
+ * Name: pm_metrics_domain
+ *
+ * Description:
+ *   This function is called when new domain got registered during pm_monitoring
+ *   or during pm_metrics initialization. It initialize the PM Metrics for given
+ *   domain.
+ *
+ * Input parameters:
+ *   domain_id - the ID of domain registered with PM.
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_metrics_domain(int domain_id)
+{
+	if (g_pm_metrics) {
+		g_pm_metrics->dtime[domain_id] = clock_systimer();
+		g_pm_metrics->suspend_ticks[domain_id] = 0;
+	}
 }
 
-void pm_prune_history(sq_queue_t *q)
+/****************************************************************************
+ * Name: pm_metrics_suspend
+ *
+ * Description:
+ *   This function is called inside pm_suspend. It note the timestamp (in ticks) of
+ *   suspended domain.
+ *
+ * Input parameters:
+ *   domain_id - the ID of domain registered with PM.
+ *
+ * Returned value:
+ *   None
+ *
+ ****************************************************************************/
+void pm_metrics_suspend(int domain_id)
 {
-	sq_queue_t *queue = q;
-	struct pm_statechange_s *sc = NULL;
-	struct timespec ts;
-	time_t cur_time;
-	clock_gettime(CLOCK_REALTIME, &ts);
-	cur_time = ts.tv_sec;
+	if (g_pm_metrics && (g_pmglobals.suspend_count[domain_id] == 0)) {
```

Comment:
All threads can access this at the same time.

Could you check if global variables can be accessed at the same time?
if there is this case, please add pm_lock(sem) or enter_critical_section